### ISSUE TITLE
Confused diagonal inputs in Demon's Souls

### ISSUE THREAD

**Issue by itsotter** [RPCS3/rpcs3#11262]
Video: https://www.youtube.com/watch?v=fWKEuSKZUWQ

While walking in a diagonal direction, I'm very frequently getting my roll sent the wrong way. In the video I'm holding [↑→] and getting a lot of [↑←] rolls. On my (xinput) gamepad this doesn't happen while holding [↑←] or [↓→], but does happen holding [↑→] or [↓←].  [↓←] gives [↓→] rolls.

This is not a quirk of the game itself! When I switch to keyboard controls in RPCS3, the problem remains but in the **other** directions: now [↑→] and [↓←] rolls consistently go where they're supposed to, but [↑←] and [↓→] rolls are often sending me the wrong way. So Demon's Souls is ok with holding any of these diagonals and rolling; I can swap between keyboard and gamepad to get any particular roll I want every time. But that's not very fun.

I'm using a new install of RPCS3 0.0.19-13104-43b7d1fe Alpha, all default settings except for Write Color Buffers checked (required for Demon's Souls). On updated Windows 10, Ryzen 5 3600X. I've been testing this with no patches applied, though it's present with the 60fps patch as well.

[deslog.txt](https://github.com/RPCS3/rpcs3/files/7728684/deslog.txt)

**Comment by Megamouse** [comment]
Please upload a proper log file after running the game

**Comment by Megamouse** [comment]
also, post a screenshot of your gamepad settings dialog for xinput

**Comment by Megamouse** [comment]
This seems to be a game bug, unless it's some inaccuracy in emulation, which is unrelated to the actual input though.

The game seems to be unable to handle the far corners of the stick input during evasive rolls.
The reason for this is most likely the fact that a real DS3 stick doesn't really reach a full square.

In the following example imagine the square is the full practical range of the input, ranging from 0 to 255 on both axis.
The center of the square is the resting position of the stick, some value like 128 for both x and y.
The green part is an exaggerated approximation of what the DS3 stick will be able to cover.
Keep in mind that a stick physically moves in a circular fashion, not like a square.
So while doing the evasion roll, the game probably doesn't expect the input that is inside the red corners.
Especially if it's exactly on each of the 4 corners (0,0), (0,255), (255,0) and (255,255).
![image](https://user-images.githubusercontent.com/23019877/146451043-b8e7590a-75ae-4fb0-ae89-07ba47087de0.png)


I think you can work around this issue by setting your left stick multiplier to something between 0.95 and 1

**Comment by Megamouse** [comment]
I just tried with 0.99 for the left stick and the issue was completely gone during my test.
So it's probably safe to assume that the game only freaks out for those 4 maxed corners.

**Comment by itsotter** [comment]
Thanks so much for taking a look.

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/7730847/RPCS3.log.gz)
![rpcs3pad](https://user-images.githubusercontent.com/33299517/146460279-0f853c0f-48a6-41c0-86c7-4ea9ebad0203.png)

Your explanation about the corners makes perfect sense, thank you! Unfortunately I'm still getting the errors when I set 0.99, 0.95, 0.90, or even 0.85 as the left stick multiplier. I tried going berserk and setting it all the way down to 0.65, at which point my character is walking, and then it *mostly* went away but not quite (happening maybe 5-10% of the time instead of 50%+).

I would take any suggestion about how to better emulate the PS3 controller with my device (maybe by setting an outer deadzone somehow?) to work around this issue, but I'm not familiar with any software that might be relevant. Could changing the squircle value do this within RPCS3? I can't really understand the description of that setting.

Regardless, I appreciate you testing it out and not just telling me "yea DeS has some funky rolls, get good noob!"

**Comment by Megamouse** [comment]
Well. For some reason I can now reproduce the issue with the multiplier set to 0.8 as well.
I think it's unrelated to the input logic then.

**Comment by Megamouse** [comment]
Regarding your questions:

1. By setting the multiplier to something below 1 you're basically already setting an outer deadzone.

2. The squircle value basically tries to map the circular stick area of your gamepad to the rounded square of a DS3.
This means that by increasing the value you can artificially "extend" your range all the way to the corners if your stick would otherwise not reach them. I think a value of 0 will do nothing while higher values increase the range until at some point you will always get the full range of the square.
You could theoretically achieve the same thing by increasing the multiplier, but then the regular stick movements to the left/right/up/down would be too intense.

**Comment by MSuih** [comment]
There seems to be underlying game bug which is causing this. The fact that it seems to work best on dualshock controllers and then differently on other handlers suggests that the developers barely got it to work on DS3 and even slightest changes in the controller behavior breaks it.

Maybe the solution is to fix the game with a patch instead of trying to comb through the joystick handling and iron out miniscule differences?

**Comment by orbea** [comment]
There was a similar issue in dolphin-emu with some games (i.e. Tales of Symphonia).

https://bugs.dolphin-emu.org/issues/9739

Which was fixed a while ago.

https://github.com/dolphin-emu/dolphin/pull/7628

I would imagine a similar idea could work in rpcs3?

**Comment by 2mg** [comment]
Had the same issue, reported it here https://www.reddit.com/r/rpcs3/comments/qtoxkb/demons_souls_rolling_in_wrong_direction/

I think DeS really likes original PS3 hardware, but it has some jank with rolling no matter what, from what I gathered online, I've read someone has this issue on PS5 remake too, so it's probably the game's coding no matter what, and emulation and/or non-PS3 gamepads only make it worse.

**Comment by Megamouse** [comment]
I also feel like this is some weird game bug and not really caused by us.
I personally also tried on PS5 remake but couldn't reproduce, which isn't really surprising since it's a different game.

**Comment by 2mg** [comment]
> I also feel like this is some weird game bug and not really caused by us. I personally also tried on PS5 remake but couldn't reproduce, which isn't really surprising since it's a different game.

PS5 version has 8 diagonal rolls too... Unless we can find someone with DS2/DS3 controller to test this out, we can't really know.
I do know that switching to KB+M didn't really help. I even emulated Gamepad > KB+M > RPCS3 :D

**Comment by 2mg** [comment]
Small addendum: if you press Circle (especially if holding forward) to exit the final Menu, you character will roll forward. 

This means that the game got 2x Circle inputs - one to close the menu and immediately one to roll. 
This points to a different problem than just analog stick/diagonals issue.

**Comment by 2mg** [comment]
A new "finding" is at 0.8 Multipliers - one is when holding Circle entire time and it shows that only ONE diagonal is peculiar in that it doesn't have a "stop and turn" animation when running. The other is when pressing Circle to start running each time, and again that peculiar diagonal has the wrong roll issue. This is on a Dinput gamepad ran thru x360ce, results may vary with direct Dinput in RPCS3 or with KB+M, but there is always ONE peculiar diagonal.


https://user-images.githubusercontent.com/7749968/170171673-c73c98e0-fd5a-450b-8dbc-7cf0a0598f91.mp4


https://user-images.githubusercontent.com/7749968/170171679-f42d4a87-6c39-4827-bd03-cf6c43c19027.mp4

**Comment by 2mg** [comment]
Seems the issue is also because some Dinput > Xinput apps don't send circles/squircles (rounded edges) inputs, but RPCS3 also **doesn't** apply a squircling effect on those rectangular Xinput shapes.

**Comment by 2mg** [comment]
Tested with a PS4 clone gamepad in Xinput mode from DS4Windows (RPCS3 DS4 handler wouldn't recognize the gamepad even without DS4Windows).

This gamepads natively have a Squicle sticks, and DS4Windows translates this behavior in Xinput mode too.

What RPCS3 should be able to do is when using a swapped Dpad <> Stick (or Keyboard for that matter) for movement, and apply a Squircle effect on the diagonals. It is impossible to have a keyboard with Squircle movements, and it just ends up with fully square behavior instead of rounded edges.

**Comment by 2mg** [comment]
Bump, this is still unresolved, no option to "force" squircle when using KBM, Dinput, or Dinput>Xinput wrappers.

**Comment by Megamouse** [comment]
This has already been implemented

**Comment by JalenJaceX2** [comment]
I still have this issue with Keyboard inputs

**Comment by QueenieArtemisia** [comment]
I am experiencing the same issue with a keyboard. Demon's Souls.

The diagonal dodges happen often in the opposite direction.

I tested all sorts of stick settings and it is still happening (1, 0.95, 0.99, 0.8, etc).

Has there ever been a fix to this?

**Comment by 2mg** [comment]
Perhaps the issue isn't in the gamepad at all but some underlying emulation + game engine quirk.
Or squircle still isn't applied (or the user didn't set it/properly tuned it).

I've done my best to provide info here and in my own ticket.
If people are still complaining and devs are reading this, well, seems it's still an issue.

**Comment by QueenieArtemisia** [comment]
Indeed, I have tried my absolute best to eliminate the problem (even the squircle settings). Nothing helped.

**Comment by 2mg** [comment]
> Indeed, I have tried my absolute best to eliminate the problem (even the squircle settings). Nothing helped.

Reopen the issue.

**Comment by QueenieArtemisia** [comment]
It says that I do not have the permission to do so.
